cmd/snap/cmd_snap_op.go

```diff
@@ -20,27 +20,70 @@
 package main
 
 import (
+	"fmt"
 	"strings"
 	"time"
 
 	"github.com/ubuntu-core/snappy/client"
 	"github.com/ubuntu-core/snappy/i18n"
+	"github.com/ubuntu-core/snappy/progress"
 
 	"github.com/jessevdk/go-flags"
 )
 
-func wait(client *client.Client, uuid string) error {
+func wait(client *client.Client, id string) error {
+	// FIXME: progress is all a bit simplistic, however its ok
+	//        for now because the only meaningful progress
+	//        we have is the download progress
+
+	// we may have multiple downloads in a single change
+	lastTotal := 0
+	pb := progress.NewTextProgress()
+	defer func() {
+		if lastTotal > 0 {
+			pb.Set(float64(lastTotal))
+		}
+		pb.Finished()
+	}()
+
 	for {
-		op, err := client.Operation(uuid)
+		chg, err := client.Change(id)
 		if err != nil {
 			return err
 		}
+		total := 1
+		msg := ""
+		for _, t := range chg.Tasks {
+			if t.Status == "Doing" {
+				msg := t.Summary
+				// this will break once we have multiple
+				// downloads in parallel in a single change
+				if t.Progress.Total > 1 {
+					cur := t.Progress.Done
+					total = t.Progress.Total
+					if t.Progress.Total != lastTotal {
```

Comment:
The logic here feels a bit strange.. it's making decisions based on the total value of different tasks, and potentially making bogus decisions if they happen to match, I suspect.

Here is a suggestion. It will need Task.ID, which I think should be there anyway.

```
var lastID string 
for _, t := range chg.Tasks {
        switch {
        case t.Status != "Doing":
                continue
        case t.Progress.Total == 1:
                pb.Spin(t.Summary)
        case task.ID == lastID:
                pb.Set(float64(t.Progress.Done))
        default:
                pb.Start(t.Summary, float64(t.Progress.Total))       
        }
        break
}